narwhal/primary/src/primary.rs

```diff
@@ -289,18 +289,33 @@ impl Primary {
             config
         };
 
-        let network = anemo::Network::bind(addr.clone())
-            .server_name("narwhal")
-            .private_key(network_signer.copy().private().0.to_bytes())
-            .config(anemo_config)
-            .outbound_request_layer(outbound_layer)
-            .start(service)
-            .unwrap_or_else(|_| {
-                panic!(
-                    "Address {} should be available for the primary Narwhal service",
-                    addr
-                )
-            });
+        let network;
+        let mut retries_left = 3;
+
+        loop {
+            let network_result = anemo::Network::bind(addr.clone())
+                .server_name("narwhal")
+                .private_key(network_signer.copy().private().0.to_bytes())
+                .config(anemo_config.clone())
+                .outbound_request_layer(outbound_layer.clone())
+                .start(service.clone());
+            match network_result {
+                Ok(n) => {
+                    network = n;
+                    break;
+                }
+                Err(_) => {
+                    retries_left -= 1;
```

Comment:
we should sleep in between retries - if it fails once it will almost certainly fail again if called in rapid succession.

Also, the most likely cause of this happening would be the sockets being in TIME_WAIT state. That shouldn't happen because we set SO_REUSEADDR, but if it does, we should retry for a total of at least 60 seconds, which is how long TIME_WAIT takes to clear. So perhaps retry once per second, maybe 90 times?

Also this should be logging at `error!` level when this happens so that error log monitoring can notice it.

Reply:
If this is anemo, then TIME_WAIT isn't a thing on udp sockets.

Reply:
As we talked offline as well with @laura-makdah , the retry here will mostly help in case where the port hasn't be de-allocated fast enough while we try to spin up the primary again - for any other case (ex a component that can't be shutdown properly) this approach won't help

Reply:
@bmwill of course you are right